server/src/main/java/org/elasticsearch/indices/TermsLookup.java

```diff
@@ -82,7 +88,11 @@ public TermsLookup(StreamInput in) throws IOException {
 
     @Override
     public void writeTo(StreamOutput out) throws IOException {
-        out.writeString(type);
+        if (out.getVersion().onOrAfter(Version.V_7_0_0)) {
+            out.writeOptionalString(type);
+        } else {
+            out.writeString(type);
```

Comment:
My understanding is that before this change, `type` could never be `null` in this class (it is marked as final, and the constructor validates that it is not null). Perhaps it would be good for me to add an assertion with a comment here.

This also reminds me of something I was hoping to get your thoughts on. With this implementation, it is not possible to use a typeless terms lookup query in a mixed 6.x and 7.0 cluster. The error will occur when a 6.x nodes tries to decode the optional string that the 7.0 node is sending here. Does this seem reasonable in terms of behavior + error messaging (not sure if there is any precedent for such a change)?

Reply:
> The error will occur when a 6.x nodes tries to decode the optional string that the 7.0 node is sending here.

Exactly what I thought, with the exception that the problem happens then a 7.0 node sends an object with a null-type I think we should make sure this cannot happen though. If we want to keep the current decision in this PR to allow null-references for types in the POJOs we need to send some dummy type when serializing to a 6.x node. We should also add unit tests for this case.

Reply:
Got it, I understand your concern now. I had assumed that passing a dummy type would actually cause more confusion than simply erroring out. For example, say a user had an index with document type `my_type`, and a mixed-node cluster. Then a typeless terms lookup query will always succeed if the lookup document is on a 7.0 node, but show no error and not find the document if it lives on a 6.x node. This is because the type name `_doc` will be used for the 6.x lookup, and we will fail to find the lookup document without any indication as to why.

I'm now thinking it would be clearest if I explicitly checked for a null type here, and errored with the message "Typeless terms lookup queries are not supported if any node is running a version before 7.0."

Reply:
>  This is because the type name `_doc` will be used for the 6.x lookup, and we will fail to find the lookup document without any indication as to why.

Is there a chance to use the dummy "_doc" type in order to not break serialization but then infer the "right" type name when on the shard for this case? As far as i understand the types removal process all 6.x indices should only have one type left, even if it is named differently?

Reply:
Yes, that would be possible if we backported the part of https://github.com/elastic/elasticsearch/pull/35790 that applied to 'get' requests. I would prefer not to do that if possible because it is not so straightfoward to backport, especially since 6.x can contain some indices with multiple types, as it supports indices created in 5.x. I also don't think that using typeless terms lookups in a mixed-node cluster will be a common enough occurrence to warrant the extra work + complexity.

Reply:
Thanks for explaining, that is unfortunate.

Reply:
In the 36549 PR for the IndexRequest class I opted to serialize using the result of a call to `type()` which will return `_doc` if null. This made 7.x+6.x nodes happy to talk to each other.
It was important that IndexRequest kept null types in the calling client because `type==null` was the condition I used to know if a blank request should inherit a choice of type from a BulkRequest's choice of default.